### PR TITLE
Adjust eaf layer to newer eaf package

### PR DESCRIPTION
**Pull request by lesliebinbin**
- Fix leader key for eaf camera, it should be *eaf-open-camera* instead of *eaf-camera*
- Refer to a recent commit: https://github.com/emacs-eaf/emacs-application-framework/commit/044b0b8507a8d022d4e798ccfdbdf1f526fe9365, we should now require each application seperately

### CODE REVIEW

**Comment by dalanicolai** [comment]
Thanks for creating this PR. However, I think we should probably make the list of apps to load configurable somehow. So I suggest that you make the list a layer variable in `config.el` and then use that variable in place of the list in `packages.el`. Subsequently, this should also get updated in the documentation (README).

Would you like to take care of that? Then also, it is usually preferred/required, to submit a single commit ([here](https://github.com/syl20bnr/spacemacs/pull/13381#pullrequestreview-380313280) is a guide that explains how to squash with magit) and [here](https://github.com/syl20bnr/spacemacs/wiki/Beginner's-Guide-to-Contributing-a-Pull-Request-to-Spacemacs#updating-the-pull-request) is a guide about updating the PR before or after squashing).

Of course, this might be a little too much to ask if your time is limited. In that case, simply let us know your time is too limited.
Also, you can always ask for help/advice on [Spacemacs gitter](https://gitter.im/syl20bnr/spacemacs).

Anyway, thanks again!

File: layers/+tools/eaf/packages.el
Comment (lebensterben) [inline layers/+tools/eaf/packages.el:51]: You can rewrite
```elisp
(list 'foo 'bar)
```
as
```elisp
'(foo bar)
```

It's much simpler.

**Review by lebensterben — APPROVED** [review]
LGTM

File: layers/+tools/eaf/packages.el
Comment (MatthewZMD) [inline layers/+tools/eaf/packages.el:65]: The latest EAF has dropped the support to `js-video-player`, feel free to remove it.

**Comment by MatthewZMD** [comment]
On top of my comment, I noticed that the EAF repo url in spacemacs is still `manateelazycat/emacs-application-framework`.

The EAF repos have merged to their own Github org: [emacs-eaf](https://github.com/emacs-eaf).

Although the old url will still work and redirect to the new one, I'd suggest, as a good practice, to set the remote to `https://github.com/emacs-eaf/emacs-application-framework.git` instead.

See more [here](https://github.com/emacs-eaf/emacs-application-framework/discussions/527?sort=new#discussioncomment-1160744)

File: layers/+tools/eaf/packages.el
Comment (lesliebinbin) [reply layers/+tools/eaf/packages.el:51]: Thank you so much, I will modify based on this, it is valuable for me as elisp rookie 😄

**Review by lebensterben — CHANGES_REQUESTED** [review]
Please make changes according to https://github.com/syl20bnr/spacemacs/pull/14992#issuecomment-899824291

File: layers/+tools/eaf/packages.el
Comment (lebensterben) [inline layers/+tools/eaf/packages.el:39]: trailing `)` should not take a new line.

File: layers/+tools/eaf/packages.el
Comment (lebensterben) [inline layers/+tools/eaf/packages.el:35]: there's no point of having `with-eval-after-load 'foo` in `:init` sectiion of `(use-package foo ..)` form.

move it to `:config` and remove `with-eval-after-load`

File: layers/+tools/eaf/config.el
Comment (lebensterben) [inline layers/+tools/eaf/config.el]: Multiple issues:

1. For `defcustom`, the second positional arg is the default value and the third positional arg is the docstring.
2. The keyword arg `:type` has an invalid value `(symbol)`. At very least, it should be `'symbol`. Here you should use `'(set (const eaf-jupyter) (const eaf-browser) ...)`. See https://www.gnu.org/software/emacs/manual/html_node/elisp/Simple-Types.html
3. If you intended to make this a custom variable, you should not use `defcustom` directly.
Instead you should use `spacemacs|defc`, which is a wrapper of `defcustom`.
4. A better docstring is required. And you also need to document it in layer's README.org.

File: layers/+tools/eaf/config.el
Comment (lebensterben) [inline layers/+tools/eaf/config.el:1]: 1. It's not `packages.el`
2. The copyright year of this file is 2021.

**Comment by lesliebinbin** [comment]
> On top of my comment, I noticed that:
> 
> 1. the EAF repo url in spacemacs is still `manateelazycat/emacs-application-framework`.
> 
> The EAF repos have merged to their own Github org: [emacs-eaf](https://github.com/emacs-eaf).
> 
> Although the old url will still work and redirect to the new one, I'd suggest, as a good practice, to set the remote to `https://github.com/emacs-eaf/emacs-application-framework.git` instead.
> 
> See more [here](https://github.com/emacs-eaf/emacs-application-framework/discussions/527?sort=new#discussioncomment-1160744)
> 
> 1. These are no longer EAF dependecies, as EAF will maintain its own fork of them from now on.
> 
> ```
> (defun eaf/init-ctable ()
>   (use-package ctable))
> (defun eaf/init-deferred ()
>   (use-package deferred))
> (defun eaf/init-epc ()
>   (use-package epc))
> ;; (defun eaf/init-s ()
> ;;   (use-package s))
> ```
> 
> These lines can be safely deleted.
> 
> See the conversation [here](https://github.com/emacs-eaf/emacs-application-framework/issues/757#issuecomment-894810710) for details.
> 
> 1. All `eaf-setq` needs to change to `setq` as per [here](https://github.com/emacs-eaf/emacs-application-framework/discussions/527?sort=new#discussioncomment-1094498)
> 2. A number of keybindings declared maybe should be set only if the apps are present.

But if we remove this, the macos user might got error, since the spacemacs  in macos still use the eaf release in May

**Review by lebensterben — CHANGES_REQUESTED** [review]

**Comment by lebensterben** [comment]
Using the snippet from https://github.com/emacs-eaf/emacs-application-framework#1-download-eaf, won't correctly install `eaf`:

```elisp
(quelpa '(eaf :fetcher github
              :repo  "emacs-eaf/emacs-application-framework"
              :files ("*")))
```

It shows
```txt
Compiling file /home/lucius/.emacs.d/elpa/27.2/develop/eaf-20210821.113/eaf.el at Sat Aug 21 01:34:44 2021
Entering directory ‘/home/lucius/.emacs.d/elpa/27.2/develop/eaf-20210821.113/’
eaf.el:167:1:Error: Cannot open load file: No such file or directory, eaf-epc
Leaving directory ‘/home/lucius/.emacs.d/elpa/27.2/develop/eaf-20210821.113/’
```

File: layers/+tools/eaf/README.org
Comment (lebensterben) [inline layers/+tools/eaf/README.org:38]: please update all the links since the repository has been moved

**Comment by lebensterben** [comment]
I won't merge this PR until it's verified to be installable.

Currently it's broken.

**Comment by MatthewZMD** [comment]
> But if we remove these, the macos user might get errors, since the spacemacs in macos still use the eaf release in May

Is it possible to update the release?

On the other hand, we're actually planning to do real releases in the near future, after things are kinda more settled.

**Comment by MatthewZMD** [comment]
> Using the snippet from https://github.com/emacs-eaf/emacs-application-framework#1-download-eaf, won't correctly install `eaf`:
> 
> ```emacs-lisp
> (quelpa '(eaf :fetcher github
>               :repo  "emacs-eaf/emacs-application-framework"
>               :files ("*")))
> ```
> 
> It shows
> 
> ```
> Compiling file /home/lucius/.emacs.d/elpa/27.2/develop/eaf-20210821.113/eaf.el at Sat Aug 21 01:34:44 2021
> Entering directory ‘/home/lucius/.emacs.d/elpa/27.2/develop/eaf-20210821.113/’
> eaf.el:167:1:Error: Cannot open load file: No such file or directory, eaf-epc
> Leaving directory ‘/home/lucius/.emacs.d/elpa/27.2/develop/eaf-20210821.113/’
> ```

Ouch, this quelpa recipe is from the pre-split period not so long ago, we haven't updated it. But I only use `use-package`, not so sure what's the best way to write up a recipe at the moment, it is still an [ongoing discussion](https://github.com/emacs-eaf/emacs-application-framework/discussions/799#discussioncomment-1213012).

The `eaf-epc.el` is located in `eaf/core` directory, do you know why isn't it loaded in the `load-path`

File: layers/+tools/eaf/README.org
Comment (MatthewZMD) [inline layers/+tools/eaf/README.org:64]: Is `=eaf=app=` a typo?

Comment (lesliebinbin) [reply layers/+tools/eaf/README.org:64]: Yep, thx :smile:

**Comment by lesliebinbin** [comment]
> > Using the snippet from https://github.com/emacs-eaf/emacs-application-framework#1-download-eaf, won't correctly install `eaf`:
> > ```emacs-lisp
> > (quelpa '(eaf :fetcher github
> >               :repo  "emacs-eaf/emacs-application-framework"
> >               :files ("*")))
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > It shows
> > ```
> > Compiling file /home/lucius/.emacs.d/elpa/27.2/develop/eaf-20210821.113/eaf.el at Sat Aug 21 01:34:44 2021
> > Entering directory ‘/home/lucius/.emacs.d/elpa/27.2/develop/eaf-20210821.113/’
> > eaf.el:167:1:Error: Cannot open load file: No such file or directory, eaf-epc
> > Leaving directory ‘/home/lucius/.emacs.d/elpa/27.2/develop/eaf-20210821.113/’
> > ```
> 
> Ouch, this quelpa recipe is from the pre-split period not so long ago, we haven't updated it, I've removed from the README until a working recipe is emerged. But I only use `use-package`, not so sure what's the best way to write up a recipe at the moment, it is still an [ongoing discussion](https://github.com/emacs-eaf/emacs-application-framework/discussions/799#discussioncomment-1213012).
> 
> The `eaf-epc.el` is located in `eaf/core` directory, do you know why isn't it loaded in the `load-path`

Not quite sure but currently I set it as a custom package and install-eaf.py works, is it because there is difference when we put
```
(defconst eaf-packages
  '(ctable
    deferred
    epc
    (eaf :location (recipe
                    :fetcher github
                    :repo  "emacs-eaf/emacs-application-framework"
                    :files ("*")))))
```
in packages.el

**Comment by lesliebinbin** [comment]
> > But if we remove these, the macos user might get errors, since the spacemacs in macos still use the eaf release in May
> 
> Is it possible to update the release?
> 
> On the other hand, we're actually planning to do real releases in the near future, after things are kinda more settled.

I will test a bit on macos, use this as a custom layer first, the difference is that when the os is not linux, it will use pip alternatives and pyqt5 install from pip got issues for long time

**Comment by lebensterben** [comment]
@lesliebinbin 
Try removing eaf* from ~/.emacs.d/elpa/27.2/develop/ first

**Comment by lesliebinbin** [comment]
> @lesliebinbin
> Try removing eaf* from ~/.emacs.d/elpa/27.2/develop/ first

Test it on ubuntu 20,

Screenshot from 2021-08-21 17-44-33

Screenshot from 2021-08-21 17-45-46

**Comment by lesliebinbin** [comment]
> @lesliebinbin
> Try removing eaf* from ~/.emacs.d/elpa/27.2/develop/ first

Oh sorry, I may find some different, in my init.el file for spacemacs I use babel load and it requires the line (require 'eaf), is that makes the diffrence, if that is the case, is that a way to eager load eaf in packages.el?

Screenshot from 2021-08-21 18-03-02

**Comment by lesliebinbin** [comment]
> > Using the snippet from https://github.com/emacs-eaf/emacs-application-framework#1-download-eaf, won't correctly install `eaf`:
> > ```emacs-lisp
> > (quelpa '(eaf :fetcher github
> >               :repo  "emacs-eaf/emacs-application-framework"
> >               :files ("*")))
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > It shows
> > ```
> > Compiling file /home/lucius/.emacs.d/elpa/27.2/develop/eaf-20210821.113/eaf.el at Sat Aug 21 01:34:44 2021
> > Entering directory ‘/home/lucius/.emacs.d/elpa/27.2/develop/eaf-20210821.113/’
> > eaf.el:167:1:Error: Cannot open load file: No such file or directory, eaf-epc
> > Leaving directory ‘/home/lucius/.emacs.d/elpa/27.2/develop/eaf-20210821.113/’
> > ```
> 
> Ouch, this quelpa recipe is from the pre-split period not so long ago, we haven't updated it, I've removed from the README until a working recipe is emerged. But I only use `use-package`, not so sure what's the best way to write up a recipe at the moment, it is still an [ongoing discussion](https://github.com/emacs-eaf/emacs-application-framework/discussions/799#discussioncomment-1213012).
> 
> The `eaf-epc.el` is located in `eaf/core` directory, do you know why isn't it loaded in the `load-path`

I thinks it is because we use defer t in packages.el for eaf, that disables the autoloads facility for eaf, I remove the defer

**Comment by lebensterben** [comment]
you're mixing things altogether.

quelpa downloads the package and install it. use-package loads the package.

Not related

**Comment by lebensterben** [comment]
@lesliebinbin 
The "do not merge" label simply means there's a hard error, which I described in earlier comments.
It's set to notify other maintainers that at current stage merging this PR causes issues.

It doesn't mean this PR is out of consideration.

**Comment by dalanicolai** [comment]
> But if we remove these, the macos user might get errors, since the spacemacs in macos still use the eaf release in May

@lesliebinbin I am not sure how Spacemacs works on macos, but I guess you are
using the same configuration as anyone else. So probably you have to update the
packages from the Spacemacs startup page (or press `SPC f e U`)

**Comment by dalanicolai** [comment]
@lesliebinbin Thanks for the work. The PR looks almost good to me.
@lebensterben @MatthewZMD Also thanks for your work guys.

I have tested this PR on a clean Spacemacs install, and it seems to work fine on
Fedora (after applying a small fix to the install-eaf.el script
https://github.com/emacs-eaf/emacs-application-framework/pull/841).

**Comment by dalanicolai** [comment]
> Using the snippet from https://github.com/emacs-eaf/emacs-application-framework#1-download-eaf, won't correctly install `eaf`:
> 
> ```emacs-lisp
> (quelpa '(eaf :fetcher github
>               :repo  "emacs-eaf/emacs-application-framework"
>               :files ("*")))
> ```
> 
> It shows
> 
> ```
> Compiling file /home/lucius/.emacs.d/elpa/27.2/develop/eaf-20210821.113/eaf.el at Sat Aug 21 01:34:44 2021
> Entering directory ‘/home/lucius/.emacs.d/elpa/27.2/develop/eaf-20210821.113/’
> eaf.el:167:1:Error: Cannot open load file: No such file or directory, eaf-epc
> Leaving directory ‘/home/lucius/.emacs.d/elpa/27.2/develop/eaf-20210821.113/’
> ```

I think you can simply change this to:
```
(quelpa '(eaf :fetcher github
              :repo  "emacs-eaf/emacs-application-framework"
              :files ("*" "core/*.el")))
```

**Comment by lebensterben** [comment]
yes this works.
```
(quelpa '(eaf :fetcher github
              :repo  "emacs-eaf/emacs-application-framework"
              :files ("*" "core/*.el")))
```

**Comment by MatthewZMD** [comment]
>               :files ("*" "core/*.el")))

I think `extension/*.el` will also be needed.

Since it works for you, could you please check this [discussion](https://github.com/emacs-eaf/emacs-application-framework/discussions/799#discussioncomment-1205285) which reported on straight, I want to know does quelpa have the same behavior - which copies the contents in the `app` directory to a new destination?

**Comment by lebensterben** [comment]
quelpa clone the whole repository.
that thread is too long and I don't want to read.

**Comment by dalanicolai** [comment]
> > ```
> >           :files ("*" "core/*.el")))
> > ```
> 
> I think `extension/*.el` will also be needed.
> 
> Since it works for you, could you please check this [discussion](https://github.com/emacs-eaf/emacs-application-framework/discussions/799#discussioncomment-1205285) which reported on straight, I want to know does quelpa have the same behavior - which copies the contents in the `app` directory to a new destination?

Quelpa builds the package in the same directory as the source. But I had a look at the straight problem. I will comment in that discussion.

**Pull request closed without merging**
